### PR TITLE
[SW-2415] Generate Algorithm-specific MOJO Scala Classes

### PR DESCRIPTION
**Pull request by mn-mikke**
Apologizes for a bigger PR, but tried to do it minimalistic as much as possible and also not to break anything.

### CODE REVIEW

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:50]: The most of the fields will have to be ignored directly on algorithms since their existence on particular algorithms doesn't make much sense. This is a consequence of generating SW API from generic V3 classes from H2O-3.

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:86]: @maurever Please could help me to identify parameters that doesn't make sense for KMeans?

Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:86]: (I mean from the below list)

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:84]: @koniecsveta   Please could help me to identify parameters from the list that doesn't make any sense for Deep Learning?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:73]: @Pscheidl @wendycwong  Please could help me to identify parameters from the list that doesn't make any sense for GAM?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:101]: @wendycwong  Please could help me to identify parameters from the list that doesn't make any sense for GLRM?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:63]: @wendycwong  Please could help me to identify parameters from the list that doesn't make any sense for GLM?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:56]: @honzasterba Please could help me to identify parameters from the list that doesn't make any sense for XGBoost?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (maurever) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:86]: The list looks correct based on the documentation of Kmeans.

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (koniecsveta) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:84]: "gainslift_bins", "custom_metric_func", "max_categorical_levels", "custom_distribution_func" doesn't make sense based on the documentation for DL.

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala
Comment (jakubhava) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:249]: just question, H2O is deprecating the model json?

File: scoring/src/main/scala/ai/h2o/sparkling/ml/params/HasMonotoneConstraintsOnMOJO.scala
Comment (jakubhava) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/params/HasMonotoneConstraintsOnMOJO.scala:54]: is it possible to set pass `monotoneConstraints` instead of string?

**Review by jakubhava — APPROVED** [review]
LGTM, from the SW perspective ( not sure about the extra algorithm params, for that we need to wait for reviews from other folks), nice change.

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala
Comment (mn-mikke) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:249]: Nope, this is just a comment for me to deprecate `trainingParams`. Since we have algorithm specific MOJO classes with a getter method for each parameter, `trainingParams` are not needed IMO. I have a plan to do the deprecation in a separate PR.

File: scoring/src/main/scala/ai/h2o/sparkling/ml/params/HasMonotoneConstraintsOnMOJO.scala
Comment (mn-mikke) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/params/HasMonotoneConstraintsOnMOJO.scala:54]: Yes, good suggestion. thanks!

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala
Comment (jakubhava) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:249]: Thanks for the explanation!

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (honzasterba) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:56]: please elaborate, parameters for what list doesnt make any sense for what?

**Review by maurever — APPROVED** [review]

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala
Comment (honzasterba) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:56]: why not use XGBoostParametersV3.fields?

Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:56]: @honzasterba That's a good point! I've exposed all fields with API notation from the class and parent classes. Just want to confirm, the `XGBoostParametersV3.fields`, `GBMParametersV3.fields`, `...` contains the list of relevant parameters for the particular algorithm and there are no exceptions. Is that correct?

**Pull request merged**
